### PR TITLE
fix: support free-threaded CPython with GIL disabled

### PR DESCRIPTION
**Pull request by colesbury**
## Description

Accesses to the `internals` struct  now goes through a `with_internals([] { ... });` statement. In the free-threaded build, the internals mutex is locked for the duration of the access.

The `registered_instances` map has its own locking scheme because it may be frequently accesses by multiple thread. The map is shareded by address and a lock per shard is used to increase concurrency.

Other smaller changes:
* Add `gil_not_used()` tag to indicate that a module supports running with the GIL disabled. See [`PyUnstable_Module_SetGIL`](https://docs.python.org/3.13/c-api/module.html#c.PyUnstable_Module_SetGIL).
* Use `PyMem_MALLOC` for `tp_doc` in 3.13+
* Add `dict_getitemstringref` to wrap new `PyDict_GetItemStringRef` API

## Suggested changelog entry:

```rst
* Support free-threaded CPython (3.13t)
* Add `py::mod_gil_not_used()`` tag to indicate if a module supports running with the GIL disabled.
```

### CODE REVIEW

File: include/pybind11/detail/internals.h
Comment (henryiii) [inline include/pybind11/detail/internals.h:189]: Would it make sense to protect this with `ifdef Py_GIL_DISABLED?` A free-threaded build can't talk to a non-free-threaded build? And this isn't needed for non-free-threaded builds. (Not sure what's planned for stable ABI going forward, don't want to create an issue if ABI4 supports both)

File: tests/test_cmake_build/main.cpp
Comment (henryiii) [inline tests/test_cmake_build/main.cpp:5]: General question: is there an inverse? Is the there some way to say "I'm wrapping a library that is single threaded (like an old C library) so don't warn about requiring the GIL in free-threaded builds (and keep the GIL even if GIL=0), or should users manually lock?

File: include/pybind11/pybind11.h
Comment (henryiii) [inline include/pybind11/pybind11.h:1307]: Should we look at switching to multiphase initialization? (if so, not sure how this would correspond)

File: include/pybind11/detail/internals.h
Comment (colesbury) [reply include/pybind11/detail/internals.h:189]: Yeah, this can just be guarded by `Py_GIL_DISABLED`. You are right that a free-threaded build can't talk to a non-free-threaded build.

I had it this way just because it seemed "cleaner" in a sense to have the fields determined solely by `PYBIND11_INTERNALS_VERSION` instead of both `PYBIND11_INTERNALS_VERSION` and `Py_GIL_DISABLED`, but either way should work.

Comment (henryiii) [reply include/pybind11/detail/internals.h:189]: It would allow us to continue to increment `PYBIND11_INTERNALS_VERSION` for GIL builds. I'll see what others think, and I can change it if others agree.

File: tests/test_cmake_build/main.cpp
Comment (colesbury) [reply tests/test_cmake_build/main.cpp:5]: We can change `set_gil_not_used()` to take an argument, but there isn't really an inverse. [`PyUnstable_Module_SetGIL`](https://docs.python.org/3.13/c-api/module.html#c.PyUnstable_Module_SetGIL) takes two options: [`Py_MOD_GIL_USED`](https://docs.python.org/3.13/c-api/module.html#c.Py_mod_gil.Py_MOD_GIL_USED) and [`Py_MOD_GIL_NOT_USED`](https://docs.python.org/3.13/c-api/module.html#c.Py_mod_gil.Py_MOD_GIL_NOT_USED), but passing `Py_MOD_GIL_USED` has the same effect as not calling the function all.

In other words, there's no way to say "keep the GIL even if `PYTHON_GIL=0`". The environment variable always overrides the module behavior.

File: include/pybind11/pybind11.h
Comment (colesbury) [reply include/pybind11/pybind11.h:1307]: It's not really relevant for free-threading, but it would allow you to better support multiple interpreters.

Comment (colesbury) [reply include/pybind11/pybind11.h:1307]: The multi-phase init equivalent is a [`PyModuleDef_Slot`](https://docs.python.org/3/c-api/module.html#c.PyModuleDef.m_slots) with `{Py_mod_gil, Py_MOD_GIL_NOT_USED}`. For example:

https://github.com/python/cpython/blob/df93f5d4bf9d70036d485666d4dd4f009d37f8b9/Modules/_bisectmodule.c#L465

Comment (henryiii) [reply include/pybind11/pybind11.h:1307]: My thought is the API, I'm not sure if `m.set_gil_not_used()` would translate to multiphase initialization?

Comment (colesbury) [reply include/pybind11/pybind11.h:1307]: It would not directly translate because you would want to set the flag on the `PyModuleDef` instead of the `PyModuleObject`, so you'd probably want to support it in `create_extension_module()`.

File: tests/test_cmake_build/main.cpp
Comment (henryiii) [reply tests/test_cmake_build/main.cpp:5]: No need to change it unless there was a way to set it without the warning.

File: include/pybind11/pybind11.h
Comment (henryiii) [reply include/pybind11/pybind11.h:1307]: I'm thinking we could add options to `create_extension_module`, which then could be passed through `PYBIND11_MODULE`, so this could become `PYBIND11_MODULE(name, m, py::gil_not_used())` for example. This would be compatible with multiphase init.

Comment (colesbury) [reply include/pybind11/pybind11.h:1307]: Yeah, I think that would be nicer. I'd imagine you'd want the new argument to be optional -- I'm not exactly sure how to do that with macros. Maybe something like:

```
#define PYBIND11_MODULE(name, variable, ...)
```

and then do something with `__VA_ARGS__`.

File: tests/test_kwargs_and_defaults.py
Comment (rwgk) [inline tests/test_kwargs_and_defaults.py:385]: "Changes an existing test" stood out to me when glancing through quickly (I can look carefully only later).

Why does this test (and test_pytypes.py) need to be changed?

Comment (colesbury) [reply tests/test_kwargs_and_defaults.py:385]: The free-threaded build immortalizes most constants, including numbers and byte literals, so they don't work well for refcount tests.

Here, `object()` is a good stand in for something that's refcounted. In `test_pytypes.py`, passing the literal through `str.encode` gives a non immortalized `bytes` object.

Comment (henryiii) [reply tests/test_kwargs_and_defaults.py:385]: I can pull out this change, and also change the check afterwards which includes the immortal object value. This is a more interesting check with `object()`, since you can't track increment and decrement if something is immortal.

Comment (henryiii) [reply tests/test_kwargs_and_defaults.py:385]: Done in #5150.

File: include/pybind11/pybind11.h
Comment (henryiii) [reply include/pybind11/pybind11.h:1307]: Yes, I do that in CLI11: https://github.com/CLIUtils/CLI11/blob/08d840bfbe92c517c279c7527323d0d1d4470e2c/include/CLI/App.hpp#L39-L46.

Hardest part is setting up the `py::gil_not_used()` handling, we do it all over the place though, so should be easy. Since there's only one option currently it is also easy to fake it with two different signatures. I can take a stab at it soon. Is it find to add the slot without adding the init/exec slots? I don't mind using the single phase method internally for now too, I just want the API to support multiphase if we add it eventually.

Comment (henryiii) [reply include/pybind11/pybind11.h]: I added a commit that does this, using the simple path for now, template meta programming can wait until we have more than one tag (unless someone wants to volunteer to do it!).

Comment (henryiii) [reply include/pybind11/pybind11.h]: (I’ll pull this commit off tomorrow and make a PR to this PR instead)

Comment (henryiii) [reply include/pybind11/pybind11.h:1307]: PR at https://github.com/colesbury/pybind11/pull/11. CI needs approval.

File: include/pybind11/detail/internals.h
Comment (wjakob) [inline include/pybind11/detail/internals.h:40]: Rather than making this a purely version-specific thing, I would rather append something to the version identifier when performing a build for the free-threaded version.

File: include/pybind11/detail/internals.h
Comment (wjakob) [inline include/pybind11/detail/internals.h:181]: maybe `alignas()` for the struct instead? That also adds padding, and then you get alignment on top.

File: include/pybind11/detail/internals.h
Comment (wjakob) [inline include/pybind11/detail/internals.h:199]: I assume the ABI would remain unchanged in a non-free-threaded build? It's not 100% clear just from looking at this change.

File: include/pybind11/detail/internals.h
Comment (wjakob) [inline include/pybind11/detail/internals.h:684]: If you don't hash the whole pointer, my suggestion would be to use a cheaper 32 bit hash.

File: include/pybind11/pybind11.h
Comment (wjakob) [inline include/pybind11/pybind11.h:2600]: hmm, this will have a non-negligible cost. You may at least want to #ifdef-gate it to the free-threaded case.

File: include/pybind11/pybind11.h
Comment (henryiii) [reply include/pybind11/pybind11.h:1209]: I like how much simpler it is. I'm not sure about the `mod`; if we put anything, it should have the full `module`, we don't use `mod` elsewhere for Python, and it isn't clearly module. But it can't be misplaced, and we don't "namespace" anything else like it.

I'm not sure it would actually be any simpler if we added a second option later (like some sort of multiphase init option). But maybe it's better to simplify now rather than worry too much about a future (which would not be more complex by this style either, at least).

`py::free_threaded()` seems like it would match the naming of CPython better, though most of the CPython internals related to the feature seem to refer to the GIL.

Comment (rwgk) [reply include/pybind11/pybind11.h:1209]: > I like how much simpler it is.

It's simpler only here, but my suggested alternative is simpler overall, because it avoids the overloads and the `private` implementation of the `create_extension_module` function. For the user it doesn't make a difference.

My suggestion includes the `mod_` so that it comes out very similar to the existing name: `Py_MOD_GIL_NOT_USED` vs. py::mod_gil_not_used().

Another thought was that it's very easy to experiment, e.g. if

```
PYBIND11_MODULE(external_module, m, py::mod_gil_not_used())
```

has some issues that are difficult to resolve in the short term, it can be changed to

```
PYBIND11_MODULE(external_module, m, py::mod_gil_not_used(false))
```

and it'll be very obvious that free-threading support was attempted, and how to start trying again. It'll also be very easy to make the free-threading support platform-specific this way.

Comment (henryiii) [reply include/pybind11/pybind11.h:1209]: > but my suggested alternative is simpler overall, because it avoids the overloads and the private implementation

That's what I meant by simpler.

Comment (henryiii) [reply include/pybind11/pybind11.h:1209]: It's also called `PyUnstable_Module_SetGIL`, which does have the `module` in it. But I hadn't noticed how that mimics the slot value name.

Comment (rwgk) [reply include/pybind11/pybind11.h:1209]: > That's what I meant by simpler.

Oh, got it. Thanks. (The way it appeared in the email message got me on the wrong track.)

> And `py::free_threaded(false)` avoids the double negative.

That works for me, too. — Usually I strive to not introduce new names if there is one already, but the double negative bothered me, too.

Comment (henryiii) [reply include/pybind11/pybind11.h:1209]: Happy to have @colesbury have final say on what name he prefers.

**Review by colesbury — COMMENTED** [review]
Thanks for the review @rwgk. I've added some comments and renamed a few functions.

**Review by henryiii — APPROVED** [review]
@colesbury If you could let us know if you like @rwgk's variation (I don't mind applying it if you do), otherwise it looks good to me.

**Comment by henryiii** [comment]
Changelog in the description is a little out of date, too.

**Comment by colesbury** [comment]
Yes, please apply it

File: include/pybind11/pybind11.h
Comment (henryiii) [inline include/pybind11/pybind11.h:1212]:
```suggestion
    explicit mod_gil_not_used(bool flag = true) : flag_(flag) {}
```

**Review by rwgk — APPROVED** [review]
Thanks @colesbury for transporting pybind11 into the free-threading world :-)

**Pull request merged**
